src/category_theory/triangulated/pretriangulated.lean

```diff
@@ -0,0 +1,219 @@
+(distinguished_triangles : set(triangle C))
+(isomorphic_distinguished : Π (T₁ ∈ distinguished_triangles) (T₂ : triangle C) (T₁ ≅ T₂),
+  T₂ ∈ distinguished_triangles)
+(contractible_distinguished : Π (X : C), (contractible_triangle C X) ∈ distinguished_triangles)
+(distinguished_cocone_triangle : Π (X:C) (Y:C) (f: X ⟶ Y), (∃ (Z : C) (g : Y ⟶ Z) (h : Z ⟶ X⟦1⟧),
+  ({ obj₁ := X, obj₂ := Y, obj₃ := Z, mor₁ := f, mor₂ := g, mor₃ := h} : triangle C) ∈
+    distinguished_triangles))
+(rotate_distinguished_triangle : Π (T : triangle C),
+  T ∈ distinguished_triangles ↔ T.rotate C ∈ distinguished_triangles)
+(complete_distinguished_triangle_morphism : Π (T₁ T₂ : triangle C)
+  (h₁ :T₁ ∈ distinguished_triangles) (h₂ :T₂ ∈ distinguished_triangles) (a : T₁.obj₁ ⟶ T₂.obj₁)
+  (b : T₁.obj₂ ⟶ T₂.obj₂) (comm₁ : T₁.mor₁ ≫ b = a ≫ T₂.mor₁),
+  (∃ (c : T₁.obj₃ ⟶ T₂.obj₃), (T₁.mor₂ ≫ c = b ≫ T₂.mor₂) ∧ (T₁.mor₃ ≫ a⟦1⟧' = c ≫ T₂.mor₃) ))
+
+/--
+Given any distinguished triangle `T`, then we know `T.rotate C` is also distinguished.
+-/
+lemma rot_of_dist_triangle [CT : pretriangulated C] (T ∈ CT.distinguished_triangles) :
+  (T.rotate C ∈ CT.distinguished_triangles) :=
+by exact (pretriangulated.rotate_distinguished_triangle T).mp H
+
+/--
+Given any distinguished triangle `T`, then we know `T.inv_rotate C` is also distinguished.
+-/
+lemma inv_rot_of_dist_triangle [CT : pretriangulated C] (T ∈ CT.distinguished_triangles) :
+  (T.inv_rotate C ∈ CT.distinguished_triangles) :=
+begin
+  apply (pretriangulated.rotate_distinguished_triangle (T.inv_rotate C)).mpr,
+  exact pretriangulated.isomorphic_distinguished T H (triangulated.triangle.rotate C
+    (triangulated.triangle.inv_rotate C T)) T ((inv_rot_comp_rot C).symm.app T),
+end
+
+/--
+Given any distinguished triangle
+```
+      f       g       h
+  X  ───> Y  ───> Z  ───> X⟦1⟧
+```
+the composition `f ≫ g = 0`.
+See https://stacks.math.columbia.edu/tag/0146
+-/
+lemma comp_dist_triangle_mor_zero₁₂ [CT : pretriangulated C] (T ∈ CT.distinguished_triangles) :
+  T.mor₁ ≫ T.mor₂ = 0 :=
+begin
+  have h := pretriangulated.contractible_distinguished T.obj₁,
+  have f := CT.complete_distinguished_triangle_morphism,
+  specialize f (contractible_triangle C T.obj₁) T h H (𝟙 T.obj₁) T.mor₁,
+  have t : (triangulated.contractible_triangle C T.obj₁).mor₁ ≫ T.mor₁ = 𝟙 T.obj₁ ≫ T.mor₁,
+    by refl,
+  specialize f t,
+  cases f with c f,
+  simp at c,
+  rw ← f.left,
+  simp only [limits.zero_comp, triangulated.contractible_triangle_mor₂],
+end -- TODO : tidy this proof up
+
+/--
+Given any distinguished triangle
+```
+      f       g       h
+  X  ───> Y  ───> Z  ───> X⟦1⟧
+```
+the composition `g ≫ h = 0`.
+See https://stacks.math.columbia.edu/tag/0146
+-/
+lemma comp_dist_triangle_mor_zero₂₃  [CT : pretriangulated C] (T ∈ CT.distinguished_triangles) :
+  T.mor₂ ≫ T.mor₃ = 0 :=
+begin
+    have H₁ := rot_of_dist_triangle C T H,
+    exact comp_dist_triangle_mor_zero₁₂ C (triangulated.triangle.rotate C T) H₁
+end
```

Comment:
```suggestion
have H₁ := rot_of_dist_triangle C T H,
comp_dist_triangle_mor_zero₁₂ C (triangulated.triangle.rotate C T) H₁
```

I guess this works in term mode?

Reply:
Your suggested change doesn't work.

I could substitute the definition of H1 in where it is used to avoid tactic mode, but I think this might make the proof harder to follow.

What are your thoughts?

Reply:
The following works:
```suggestion
have H₁ : _ := rot_of_dist_triangle C T H,
comp_dist_triangle_mor_zero₁₂ C (triangulated.triangle.rotate C T) H₁
```
Personally, I like the one-line version best.